From 5a6b8004426374eeb86b88142bd763eeca7df365 Mon Sep 17 00:00:00 2001
From: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
Date: Mon, 11 May 2026 07:34:50 +0000
Subject: [PATCH] nvme-tcp: fix UAF when detecting digest errors

jira VULN-34111
cve CVE-2022-48686
commit-author Sagi Grimberg <sagi@grimberg.me>
commit 160f3549a907a50e51a8518678ba2dcf2541abea

We should also bail from the io_work loop when we set rd_enabled to true,
so we don't attempt to read data from the socket when the TCP stream is
already out-of-sync or corrupted.

Fixes: 3f2304f8c6d6 ("nvme-tcp: add NVMe over TCP host driver")
	Reported-by: Daniel Wagner <dwagner@suse.de>
	Signed-off-by: Sagi Grimberg <sagi@grimberg.me>
	Reviewed-by: Daniel Wagner <dwagner@suse.de>
	Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 160f3549a907a50e51a8518678ba2dcf2541abea)
	Signed-off-by: CIQ Kernel Automation <ciq_kernel_automation@ciq.com>
---
 drivers/nvme/host/tcp.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/nvme/host/tcp.c b/drivers/nvme/host/tcp.c
index 600037cf1ac2..f6528ffec92b 100644
--- a/drivers/nvme/host/tcp.c
+++ b/drivers/nvme/host/tcp.c
@@ -1179,7 +1179,7 @@ static void nvme_tcp_io_work(struct work_struct *w)
 		else if (unlikely(result < 0))
 			return;
 
-		if (!pending)
+		if (!pending || !queue->rd_enabled)
 			return;
 
 	} while (!time_after(jiffies, deadline)); /* quota is exhausted */
-- 
2.47.3

